### PR TITLE
Libffi 3.2.1

### PR DESCRIPTION
**Pull request by s172262**
### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew/blob/master/.github/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew/pulls) for the same update/change?

### CODE REVIEW

**Comment by s172262** [comment]
9 tests were failed but 1473 tests passed.
All failed tests caused form the other formula which conflicts with another formula when install them.

File: Library/Formula/libffi.rb [hunk 1]
```diff
@@ -1,17 +1,12 @@
...
+  mirror "ftp://sourceware.org/pub/libffi/libffi-3.2.1.tar.gz"
+  sha256 "d06ebb8e1d9a22d19e38d63fdb83954253f39bedc5d46232a05645685722ca37"
 
   bottle do
     cellar :any
-    sha256 "d512d7c3258d61e088097f1f9a1fd010bd1a197e760e0b3abc08a3f767624745" => :el_capitan
-    sha256 "f75c1beb848231ed3e2275c867620da91dd16be9dc7c4b88f86675ac62159323" => :yosemite
-    sha256 "aa60d56351d36a45f2e7f16114fc17f9bd8fe805931f36d744c6ccb5fa5df238" => :mavericks
-    sha256 "fad1fe049554d37471408fe451ef2e46628177c94eaafb23a3af56336603baad" => :mountain_lion
-    sha256 "dc4718ebb77ff384386e0ef1782d8418c821637044b5dff7c08f21c401d0668d" => :lion
```
Comment (bfontaine) [inline Library/Formula/libffi.rb]: Please don’t remove these lines; our bot will handle them. Thanks!

File: Library/Formula/libffi.rb [hunk 2]
```diff
@@ -21,15 +16,35 @@ class Libffi < Formula
     depends_on "libtool" => :build
   end
 
-  keg_only :provided_by_osx, "Some formulae require a newer version of libffi."
+  keg_only :provided_by_osx," Some formulae require this version of libffi."
+
+  stable do
+    patch :DATA
```
Comment (bfontaine) [inline Library/Formula/libffi.rb]: Please add a comment above this line explaining why the patch is necessary as well as a link to the upstream’s bug report.

File: Library/Formula/libffi.rb [hunk 3]
```diff
@@ -21,15 +16,35 @@ class Libffi < Formula
...
     ENV.deparallelize # https://github.com/Homebrew/homebrew/pull/19267
-    ENV.universal_binary
     system "./autogen.sh" if build.head?
+    system "python", "./generate-darwin-source-and-headers.py", "--only-osx"
     system "./configure", "--disable-debug", "--disable-dependency-tracking",
                           "--prefix=#{prefix}"
     system "make", "install"
+    if build.universal?
+      cd("build_macosx-x86_64"){system "make"}
+      cd("build_macosx-i386"){system "make"}
```
Comment (bfontaine) [inline Library/Formula/libffi.rb]: Use `system "make", "-C", "foo"` instead of `cd("foo") { system "make" }`.

File: Library/Formula/libffi.rb [hunk 4]
```diff
@@ -21,15 +16,35 @@ class Libffi < Formula
...
+      system "lipo", "-create", "build_macosx-x86_64/.libs/libffi.6.dylib", "build_macosx-i386/.libs/libffi.6.dylib", "-output", "libffi.6.dylib"
+      system "lipo", "-create", "build_macosx-x86_64/.libs/libffi.a", "build_macosx-i386/.libs/libffi.a", "-output", "libffi.a"
+      lib.install "libffi.6.dylib"
+      lib.install "libffi.a"
```
Comment (bfontaine) [inline Library/Formula/libffi.rb]: You can do both in one line: `lib.install %w[libffi.6.dylib libffi.a]`.

File: Library/Formula/libffi.rb [hunk 5]
```diff
@@ -21,15 +16,35 @@ class Libffi < Formula
...
+    else
+      if Hardware::CPU.intel?
+        if OS::Mac.prefer_64_bit?
+          cd("build_macosx-x86_64"){system "make";lib.install ".libs/libffi.6.dylib";lib.install ".libs/libffi.a"}
+         else
+          cd("build_macosx-i386"){system "make";lib.install ".libs/libffi.6.dylib";lib.install ".libs/libffi.a"}
```
Comment (bfontaine) [inline Library/Formula/libffi.rb]: Please don’t duplicate these lines. You can use something like this:
```rb
directory = OS::Mac.prefer_64_bit? ? "build_macosx-x86_64" : "build_macosx-i386"
cd directory do
  system "make"
  lib.install %w[.libs/libffi.6.dylib .libs/libffi.a]
end
```

File: Library/Formula/libffi.rb (hunk 2 above)
Comment (s172262) [inline Library/Formula/libffi.rb]: They knew this bug and that the bug is caused form **win32.S**. But they won't accept my advice.

**Comment by apjanke** [comment]
If upstream's not taking the patch, I'm afraid we can't accept it in Homebrew.

If the situation changes, feel free to re-open on the new [`homebrew/core` repo](https://github.com/homebrew/homebrew-core).

**Comment by s172262** [comment]
Upstream say that this patch builds just fine on recent OSX versions. This doesn't build on older versions, specifically 10.6 and earlier.

**Pull request closed without merging**